src/index.ts

```diff
@@ -1,12 +1,183 @@
-import EnsemblGenomeBrowser from './ensemblGenomeBrowser';
+import { 
+  IncomingAction, IncomingActionType, OutgoingAction, OutgoingActionType
+} from './types';
 
-import { OutgoingAction, OutgoingActionType, IncomingAction, IncomingActionType } from './action';
+const subscriptions = new Map<IncomingActionType, Set<(action: IncomingAction) => void>>();
+type Callback = (action: IncomingAction) => void;
 
-export {
-    OutgoingAction, 
-    OutgoingActionType, 
-    IncomingAction, 
-    IncomingActionType
+type GenomeBrowserType = {
+  go: (config_object: any) => void;
+  set_stick: (stickId: string) => void;
+  wait: () => void;
+  goto: (left: number, right: number) => void;
+  jump: (location: string) => void;
+  set_y: (y: number) => void;
+  set_switch: (path: string[]) => void;
+  clear_switch: (path: string[]) => void;
+  set_message_reporter: (callback: (...action: [type: IncomingActionType, payload: any]) => void) => void;
+};
+
+
+type ConfigData = {
+  backend_url?: string;
+  target_element_id?: string;
+}
+class EnsemblGenomeBrowser {
+  
+  genomeBrowser: GenomeBrowserType | null = null;
+  inited = false;
+
+  public async init(config: ConfigData = {}) {
+
+    if(!this.inited) {
+      const { default: init, GenomeBrowser } = await import('./peregrine/peregrine_generic.js');
+      await init();
+      this.genomeBrowser = new GenomeBrowser();
+      this.genomeBrowser?.go(config);
+    }
+    this.inited = true;
+    
+    this.genomeBrowser?.set_message_reporter(this.handleIncoming);
+  }
+
+  public formatIncoming = (actionType: IncomingActionType, payload: any) => {
+
+    if (actionType === IncomingActionType.TRACK_SUMMARY) {
+
+        return {
+          type: actionType,
+          payload: payload.summary
+        } as IncomingAction
+    } else if (actionType === IncomingActionType.ZMENU_CREATE) {
+
+        const id = payload.content[0].metadata.transcript_id;
+        const unversioned_id = id.split('.')[0];
+        return {
+          type: actionType,
+          payload: {
+            id,
+            unversioned_id,
+            anchor_coordinates: {x: payload.x, y : payload.y},
+            content: payload.content
+          }
+        } as IncomingAction
+
+    }
+
+    return {
+      type: actionType,
+      payload
+    } as IncomingAction
+
+  }
+
+  public handleIncoming = (...action: [type: IncomingActionType, payload: any]) => {
+
+    const [type, payload] = action;
+
+    const subscriptionsToAction = subscriptions.get(type);
+
+    if (subscriptionsToAction) {
+      
+      subscriptionsToAction.forEach( subscription => {
```

Comment:
`subscriptionsToAction` is going to be a Set, which, as we discovered, can be mutated while it's going through the forEach loop. You may want to first convert it to a new array to avoid this hazard:

`[...subscriptionsToAction.values()].forEach(...)`